File: card-web-client/src/models/Creature/Hero.ts
import CardTypeEnum from '../Enum/CardTypeEnum';
import EquipmentSlotEnum from '../Enum/EquipmentSlotEnum';
import EquipmentSlot from '../Enum/EquipmentSlotEnum';
import { Effect, Equipment } from '../Useable';
import Ability from './Ability';
import IBaseCreature from './IBaseCreature';

interface IHero extends IBaseCreature {}

class Hero implements IHero {
  _id: number;
  name: string;
  description: string;
  private cardType: CardTypeEnum;
  public get type(): string {
    return CardTypeEnum[this.cardType];
  }
  public set type(value: string) {
    this.cardType = CardTypeEnum[value as keyof typeof CardTypeEnum];
  }
  imageId: string;

  level: number;
  equippedItems: Equipment[];
  equipmentSlots: EquipmentSlot[];
  effect: Effect[];
  abilities: Ability[];
  exhaust: Function;

  constructor(
    _id: number,
    name: string,
    description: string,
    cardType: CardTypeEnum,
    imageId: string,
    level: number,
    equippedItems: Equipment[],
    equipmentSlots: EquipmentSlot[],
    effect: Effect[],
    abilities: Ability[],
    exhaust: Function
  ) {
    this._id = _id;
    this.name = name;
    this.description = description;
    this.cardType = cardType;
    this.imageId = imageId;
    this.level = level;
    this.equippedItems = equippedItems;
    this.equipmentSlots = equipmentSlots;
    this.effect = effect;
    this.abilities = abilities;
    this.exhaust = exhaust;
  }
}

function createGenericHero(): Hero {
  return new Hero(
    0, // _id
    'Generic Hero Card', // name
    'This is the Hero Description', // description
    CardTypeEnum.Hero, // typeOfCard
    'genericImage.png', // image
    1, // level
    [], // equippedItems
    [
      EquipmentSlotEnum.Head,
      EquipmentSlotEnum.Arm,
      EquipmentSlotEnum.Arm,
      EquipmentSlotEnum.Torso,
      EquipmentSlotEnum.Leg,
      EquipmentSlotEnum.Leg,
      EquipmentSlotEnum.Ranged,
    ], // equipmentSlots
    [], // effect
    [], // Abilities
    () => {} // Exhaust
  );
}

export { createGenericHero };
export default Hero;

File: OLD-MML-Card-BE-DOTNET/Models/Card/Creature/Hero.cs
using MML_Card_BE.Models.Enum;
using MML_Card_BE.Models.Card.Creature.Interfaces;
using MML_Card_BE.Models.Card.Useable;
using MongoDB.Bson.Serialization.Attributes;
using MongoDB.Bson;

namespace MML_Card_BE.Models.Card.Creature
{
    public class Hero : IHero
    {
        [BsonId]
        [BsonRepresentation(BsonType.ObjectId)]
        public string? _id { get; set; }
        
        [BsonElement("name")]
        [BsonRepresentation(BsonType.String)]
        public string Name { get; set; }
        
        [BsonElement("description")]
        [BsonRepresentation(BsonType.String)]
        public string? Description { get; set; }
        
        [BsonElement("cardType")]
        [BsonRepresentation(BsonType.Int32)]
        private CardTypeEnum CardType { get; set; }
        public string type
        {
            get => CardType.ToString(); 
        }
        
        [BsonElement("imageId")]
        [BsonRepresentation(BsonType.String)]
        public string? ImageId { get; set; }

        [BsonElement("level")]
        [BsonRepresentation(BsonType.Int32)]
        public int Level { get; set; }
        
        [BsonElement("creatureType")]
        [BsonRepresentation(BsonType.Array)]
        public List<string> CreatureType { get; set; }
        
        public List<Equipment> EquippedItems { get; set; }
        
        [BsonElement("equipmentSlots")]
        [BsonRepresentation(BsonType.Array)]
        public List<EquipmentSlot> EquipmentSlots { get; set; }

        [BsonElement("effect")]
        [BsonRepresentation(BsonType.Array)]
        public List<Effect> Effect { get; set; }

        [BsonElement("abilities")]
        [BsonRepresentation(BsonType.Array)]
        public List<Ability> Abilities { get; set; }

        [BsonElement("exhaust")]
        [BsonRepresentation(BsonType.String)]
        public Func<string> Exhaust { get; set; }

        [BsonElement("cardNumber")]
        [BsonRepresentation(BsonType.Int32)]
        public int CardNumber { get; set; }

        [BsonElement("CardSet")]
        [BsonRepresentation(BsonType.String)]
        public SetEnum Set { get; set; }
    }
}
